### PR TITLE
Extrusion: Pass number of layers as argument to slate kernel.

### PR DESCRIPTION
**Pull request by sv2518**
Addresses #1715 to fix variable layers in slate code generation. What do you think about this?

### CODE REVIEW

File: firedrake/assemble.py [hunk 1]
```diff
@@ -537,6 +538,10 @@ def get_map(x):
         if needs_cell_facets:
             assert integral_type == "cell"
             extra_args.append(m.cell_to_facets(op2.READ))
+        if pass_layer_arg:
+            import numpy
+            o = base.Arg(op2.Global(1, itspace.layers-2, dtype=numpy.dtype(numpy.int32)))
```
Comment (sv2518) [inline firedrake/assemble.py:543]: I was not 100 percent sure, if this is the right place to generate the argument for the amount of layers?

Comment (wence-) [reply firedrake/assemble.py:543]: That looks reasonable. You can  just do:

```
c = op2.Global(1, itspace.layers-2, dtype=...)
o = c(op2.READ)
```
I think.

Comment (wence-) [reply firedrake/assemble.py:543]: This is a simpler fix than I was thinking of (which was to augment kernels saying that they needed this and then wrap it all up in PyOP2).

I think this approach is fine.

File: firedrake/slate/slac/compiler.py [hunk 2]
```diff
@@ -483,15 +485,15 @@ def tensor_assembly_calls(builder):
         # The mesh level is an integer provided as a macro kernel
         # argument.
 
-        # FIXME: No variable layers assumption
```
Comment (wence-) [inline firedrake/slate/slac/compiler.py:486]: This FIXME still applies (variable layers is different numbers of layers for each entry in the iteration set).

**Comment by wence-** [comment]
Mostly fine. I suspect you need to rebase on top of (or just port the patch) current master, since the merge to refactor `assemble.py` has gone in.

File: firedrake/assemble.py (hunk 1 above)
Comment (sv2518) [reply firedrake/assemble.py:619]: Your suggestion resulted in a c compiler error, so I just left it as it was.

Comment (wence-) [reply firedrake/assemble.py:619]: WAH? I don't understand, not least because your version doesn't set an access descriptor for the argument.

Comment (wence-) [reply firedrake/assemble.py:619]: Which I think should not be allowed, although I see we don't actually check.

Comment (sv2518) [reply firedrake/assemble.py:619]: Ah. I know why, I had not specified a const qualifier in the COFFEE argument declaration.

**Pull request merged**